### PR TITLE
Deconv from uit scripts

### PR DESCRIPTION
**Pull request by audunth**
Deconv_methods updates moved from uit_scripts to fpp-analysis-tools

### CODE REVIEW

File: fppanalysis/deconvolution_methods.py
Comment (gregordecristoforo) [inline fppanalysis/deconvolution_methods.py:115]: To me it seems like to a good idea to combine this function with `RL_gauss_deconvolve()` since most of the code of the gpu function is just copied from there.

**Review by gregordecristoforo — COMMENTED** [review]
To me it seems like to a good idea to combine `gpu_RL_gauss_deconvolve()` with `RL_gauss_deconvolve()` since most of the code of the gpu function is just copied from there. 
One could add a boolean parameter to switch between gpu and cpu implementation or make the gpu as a default and use the cpu if no gpu is awailable, something like:
```Python
try:
    gpu_implementation()
except:
    cpu_implementation()
```

**Comment by audunth** [comment]
We should also make sure all dependencies are covered before merging, see [#8](https://github.com/uit-cosmo/fpp-analysis-tools/issues/8).

**Comment by sajidah-ahmed** [comment]
> To me it seems like to a good idea to combine `gpu_RL_gauss_deconvolve()` with `RL_gauss_deconvolve()` since most of the code of the gpu function is just copied from there. One could add a boolean parameter to switch between gpu and cpu implementation or make the gpu as a default and use the cpu if no gpu is awailable, something like:
> 
> ```python
> try:
>     gpu_implementation()
> except:
>     cpu_implementation()
> ```

It doesn't make sense to install cuPy  and CuSignal if you are running it on a laptop without a GPU. There needs to be a check on whether to install those dependencies if you're running this on a GPU.

**Comment by gregordecristoforo** [comment]
@sajidah-ahmed can you please run a test for the new implementation of the gpu version?

File: fppanalysis/deconvolution_methods.py
Comment (sajidah-ahmed) [inline fppanalysis/deconvolution_methods.py:59]: change to xp

File: fppanalysis/deconvolution_methods.py
Comment (sajidah-ahmed) [inline fppanalysis/deconvolution_methods.py:60]: change to xp

**Review by sajidah-ahmed — CHANGES_REQUESTED** [review]

**Comment by sajidah-ahmed** [comment]
@engeir Could you please add the option for install cupy and cusignal as a poetry extras? Poetry fails to add cusignal and says the following putting down on the terminal: `poetry add cusignal`

`Failed to add packages, reverting the pyproject.toml file to its original content.`

**Comment by engeir** [comment]
@sajidah-ahmed Can look into it. I have never used `cuda` before, could you specify how you normally would install needed modules (`cupy` and `cusignal`?)? Is it with a simple `pip install cupy`, or do you need to `apt install` something as well?

**Comment by engeir** [comment]
This is the error you get? Seems like it isn't packaged for `pip`, but rather for `conda`?

<details><summary>Error log</summary>

```console
$ poetry add cusignal
Using version ^0.0.1 for cusignal

Updating dependencies
Resolving dependencies...

Package operations: 1 install, 0 updates, 0 removals

  • Installing cusignal (0.0.1)

  EnvCommandError

  Command ['/home/een023/.pyenv/versions/3.8.12/envs/fpp-analysis/bin/pip', 'install', '--no-deps', 'file:///home/een023/.cache/pypoetry/artifacts/9b/b4/0c/27ea16d94d52637000bdc64503396bfd131e6e1725d2f3fc33b4bdd1dd/cusignal-0.0.1.tar.gz'] errored with the following return code 1, and output: 
  Processing /home/een023/.cache/pypoetry/artifacts/9b/b4/0c/27ea16d94d52637000bdc64503396bfd131e6e1725d2f3fc33b4bdd1dd/cusignal-0.0.1.tar.gz
  Using legacy 'setup.py install' for cusignal, since package 'wheel' is not installed.
  Installing collected packages: cusignal
      Running setup.py install for cusignal: started
      Running setup.py install for cusignal: finished with status 'error'
      ERROR: Command errored out with exit status 1:
       command: /home/een023/.pyenv/versions/3.8.12/envs/fpp-analysis/bin/python3.8 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-j1iowcrc/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-j1iowcrc/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-wu1jtwsw/install-record.txt --single-version-externally-managed --compile --install-headers /home/een023/.pyenv/versions/3.8.12/envs/fpp-analysis/include/site/python3.8/cusignal
           cwd: /tmp/pip-req-build-j1iowcrc/
      Complete output (16 lines):
      running install
      Traceback (most recent call last):
        File "<string>", line 1, in <module>
        File "/tmp/pip-req-build-j1iowcrc/setup.py", line 18, in <module>
          setup(name=pkg,
        File "/home/een023/.pyenv/versions/3.8.12/envs/fpp-analysis/lib/python3.8/site-packages/setuptools/__init__.py", line 153, in setup
          return distutils.core.setup(**attrs)
        File "/home/een023/.pyenv/versions/3.8.12/lib/python3.8/distutils/core.py", line 148, in setup
          dist.run_commands()
        File "/home/een023/.pyenv/versions/3.8.12/lib/python3.8/distutils/dist.py", line 966, in run_commands
          self.run_command(cmd)
        File "/home/een023/.pyenv/versions/3.8.12/lib/python3.8/distutils/dist.py", line 985, in run_command
          cmd_obj.run()
        File "/tmp/pip-req-build-j1iowcrc/setup.py", line 15, in run
          raise Exception(long_description)
      Exception: Please install cusignal via the rapidsai conda channel. See https://rapids.ai/start.html for instructions.
      ----------------------------------------
  ERROR: Command errored out with exit status 1: /home/een023/.pyenv/versions/3.8.12/envs/fpp-analysis/bin/python3.8 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-j1iowcrc/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-j1iowcrc/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-wu1jtwsw/install-record.txt --single-version-externally-managed --compile --install-headers /home/een023/.pyenv/versions/3.8.12/envs/fpp-analysis/include/site/python3.8/cusignal Check the logs for full command output.
  WARNING: You are using pip version 21.1.1; however, version 22.0.4 is available.
  You should consider upgrading via the '/home/een023/.pyenv/versions/3.8.12/envs/fpp-analysis/bin/python3.8 -m pip install --upgrade pip' command.
  

  at ~/.local/share/pypoetry/venv/lib/python3.10/site-packages/poetry/utils/env.py:1195 in _run
      1191│                 output = subprocess.check_output(
      1192│                     cmd, stderr=subprocess.STDOUT, **kwargs
      1193│                 )
      1194│         except CalledProcessError as e:
    → 1195│             raise EnvCommandError(e, input=input_)
      1196│ 
      1197│         return decode(output)
      1198│ 
      1199│     def execute(self, bin, *args, **kwargs):

Failed to add packages, reverting the pyproject.toml file to its original content.
```
</details>

**Comment by sajidah-ahmed** [comment]
> This is the error you get when installing `cusignal`? Seems like it isn't packaged for `pip`, but rather for `conda`. Also, `cupy` might work ([`poetry add cupy-cuda115`](https://docs.cupy.dev/en/stable/install.html) works), as long as you have `apt` dependencies installed.
> 
> Edit: [Drops PIP packages](https://medium.com/rapids-ai/rapids-0-7-release-drops-pip-packages-47fc966e9472) and [comment in PyPI README](https://pypi.org/project/cusignal/).
> Installing `cusignal`
> Importing `cupy`

It is recommended to install `cusignal` with `conda`.

I used the following command at the time when I installed.
```
For CUDA 11.0 and Python 3.8
conda install -c rapidsai -c nvidia -c conda-forge \
    cusignal=21.08 python=3.8 cudatoolkit=11.0
```
Does `poetry` have a problem with `conda`? I honestly don't know much about `poetry` and I thought we're sticking to the usual `setup.py` files. Otherwise, it would make sense to just have a manual install of cusignal via conda as a pre-requisite if using the deconvolution on the GPUs.

**Comment by engeir** [comment]
As far as I understand, `conda` isn't just a python environment and dependency installer, but it creates an environment for the compiled binaries and so on that `cusignal`, in this case, depends on. It can even install `python` itself inside of this environment.

I think it is much easier to set up a `conda` env and install `cusignal` into that, than to do it with `pip`. Seems like installing with `pip` is actually not possible.

`poetry`, on the other hand, is "pure python", and is meant to install python modules only. There is nothing fancy or special about `poetry`, it just do what `pip` does, but in a more easy-to-control way. In conclusion, `poetry` is not really the issue in this case (`setup.py` would be equally hard, I think, since `cusignal` do not support install via PyPI).

Setting up a `conda` env from the already present `pyproject.toml` file can be done with [`poetry2conda`](https://github.com/dojeda/poetry2conda). I had to do it once when I was making plots with `cartopy`, which also has much better support in `conda` due to all the specific libraries it uses. [Here](https://github.com/engeir/cesm-helper-scripts/blob/f22b0b6b28e1cad4bd7f629018d9184f35d283e2/pyproject.toml#L53) is all I needed to add to `pyproject.toml`. I can try and make a quick commit that has an `environment.yml` file which `conda` can use.

**Comment by engeir** [comment]
Don't know if this works, but the conda env is being created, at least. (`cudatoolkit` is perhaps not needed).

`environment.yml` is created from

```sh
poetry2conda -E cupy -E cudatoolkit -E cusignal pyproject.toml environment.yml
```

and then I created a conda env with the name `conda-fpp-analysis` with command

```sh
conda env create --file environment.yml
```

**Comment by sajidah-ahmed** [comment]
> 

Okay, so I git cloned the repository, then switched branches. Which one am meant to do first? `poetry install` or
`conda env create --file environment.yml`?

I made the conda enviroment, but when I did `poetry install` it gives me an error:

```
Installing dependencies from lock file
Warning: The lock file is not up to date with the latest changes in pyproject.toml. You may be getting outdated dependencies. Run update to update them.

  SolverProblemError

  Because fppanalysis depends on cusignal (>=0.19,<=0.22) which doesn't match any versions, version solving failed.
```

It might be a good idea to update the README.md file with instructions describing in what order to do this.

**Comment by engeir** [comment]
Yes, you would do only one of them. Either
poetry install
which would not install cusignal etc., since it lives in a conda channel. 

Or you do
conda env create --file environment.yml
which creates the conda environment with the cusignal module etc.

_Running poetry install first, then conda ... might work since you probably get then get two separate environments, one that poetry use, a second that conda creates. conda first, then poetry, will cause an error since poetry will try to use the already created conda environment. I can update the README as well if all works well with the second approach of using conda._

**Comment by sajidah-ahmed** [comment]
> Yes, you would do only one of them. Either poetry install which would not install cusignal etc., since it lives in a conda channel.
> 
> Or you do conda env create --file environment.yml which creates the conda environment with the cusignal module etc.
> 
> _Running poetry install first, then conda ... might work since you probably get then get two separate environments, one that poetry use, a second that conda creates. conda first, then poetry, will cause an error since poetry will try to use the already created conda environment. I can update the README as well if all works well with the second approach of using conda._

Okay so even though I made the conda environment, `fppanalysis` is not installed as a package still and the deconvolution fails to fun since that hasn't been installed yet.

**Comment by sajidah-ahmed** [comment]
> Ahh, true. Yeah that's a problem, remember now I had that problem in my other thing I mentioned as well. And I don't know how to install the project itself in conda... So not sure what the best solution is, tbh.
> 
> One might be able to work around it by doing only relative imports, instead of `import fppanalysis`, but maybe this just cause some other issue. Not sure.

That might not be ideal but thanks for looking into this though :) I was thinking, when one does `poetry install` is there a possibility of asking a yes/no question on the terminal about running the package on the GPU - if yes: install cupy and cusignal manually; if no - nothing happens. The manual installation can be mentioned on the README file also. What do you think?

**Comment by engeir** [comment]
> Okay so even though I made the conda environment, `fppanalysis` is not installed as a package still and the deconvolution fails to run.

Hmm, I tried to do a clean install now, and it actually do work just as expected for me. Could you explain what the error you got was?

The steps I did was:

```sh
git clone ...
git checkout deconv...
conda env create --file environment.yml
conda activate conda-fpp-analysis
python
```

Then in the python interpreter:

```python
>>> import numpy as np
>>> import fppanalysis as fpp
>>> a = np.linspace(0, 100)
>>> out = fpp.RL_gauss_deconvolve(a, a, 200)
  0%|                                                                                                                                                                                                                                                          | 0/200 [00:00<?, ?it/s]i = 200
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:00<00:00, 3508.34it/s]
>>> out[0].shape
(50, 1)
```

And to verify I was in the correct python environment set up by conda:

```console
$ python -c "import sys;print(sys.executable)"
/home/een023/programs/miniconda3/envs/conda-fpp-analysis/bin/python
```

**Comment by sajidah-ahmed** [comment]
> > Okay so even though I made the conda environment, `fppanalysis` is not installed as a package still and the deconvolution fails to run.
> 
> Hmm, I tried to do a clean install now, and it actually do work just as expected for me. Could you explain what the error you got was?
> 
> The steps I did was:
> 
> ```shell
> git clone ...
> git checkout deconv...
> conda env create --file environment.yml
> conda activate conda-fpp-analysis
> python
> ```
> 
> Then in the python interpreter:
> 
> ```python
> >>> import numpy as np
> >>> import fppanalysis as fpp
> >>> a = np.linspace(0, 100)
> >>> out = fpp.RL_gauss_deconvolve(a, a, 200)
>   0%|                                                                                                                                                                                                                                                          | 0/200 [00:00<?, ?it/s]i = 200
> 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:00<00:00, 3508.34it/s]
> >>> out[0].shape
> (50, 1)
> ```
> 
> And to verify I was in the correct python environment set up by conda:
> 
> ```
> $ python -c "import sys;print(sys.executable)"
> /home/een023/programs/miniconda3/envs/conda-fpp-analysis/bin/python
> ```

The above implementation you did is fine but you're not using the GPU-implementation of the deconvolution here which needs cupy arrays instead of numpy arrays. 

I get this:

```
>>> import cupy as cp
>>> import fppanalysis as fpp
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'fppanalysis'
```
When we installed `cusignal `and `cupy `through `environment.yml,` the `fppanalysis` package wasn't installed.

**Comment by gregordecristoforo** [comment]
To me it seems it would be more practical to keep `cupy` and `cusignal` out of the `pyproject.toml` file and specify in the `README.md` that these packages must be installed manually via pip in order to use the GPU version. @sajidah-ahmed can you test whether that works with the current version of the code?

**Comment by sajidah-ahmed** [comment]
> To me it seems it would be more practical to keep `cupy` and `cusignal` out of the `pyproject.toml` file and specify in the `README.md` that these packages must be installed manually via pip in order to use the GPU version. @sajidah-ahmed can you test whether that works with the current version of the code?

@gregordecristoforo, @engeir 
I got rid of the `cupy` and `cusignal` from the `pyproject.toml` file. 
Created a new conda environment and installed `fppanalysis` via `poetry install`. 
Then did:
```
conda install -c rapidsai -c nvidia -c conda-forge \
    cusignal=21.08 python=3.8 cudatoolkit=11.0
```
Ran a script using the GPU version of the deconvolution using cupy arrays and it was all fine!

**Comment by engeir** [comment]
> @gregordecristoforo, @engeir I got rid of the `cupy` and `cusignal` from the `pyproject.toml` file. Created a new conda environment and installed `fppanalysis` via `poetry install`. Then did:
> 
> ```
> conda install -c rapidsai -c nvidia -c conda-forge \
>     cusignal=21.08 python=3.8 cudatoolkit=11.0
> ```
> 
> I ran a script using the GPU version of the deconvolution using cupy arrays and it was all fine!
> 
> Should we remove the `environment.yml` file?

There is a minor gotcha with this approach, I think.

```sh
conda install -c rapidsai -c nvidia -c conda-forge \
    cusignal=21.08 python=3.8 cudatoolkit=11.0
```

will not install the packages into any specific environment, which means that `poetry install` will install into some other environment (if `poetry` can't find one, it will create one in a new directory `.venv`) and install the project there. Thus, the stuff installed by `poetry` will be unrelated to the stuff installed by `conda`. Same happens if you run `poetry install` first.

If a `conda` environment _is already activated_, (I believe) the two commands can be run in any order you like, for example:

```sh
poetry install
conda install -c rapidsai -c nvidia -c conda-forge \
    cusignal=21.08 python=3.8 cudatoolkit=11.0
```

This is where the `environment.yml` file seems a handy tool to me, since it defines a new environment where everything should be installed. (BTW, did you try running `poetry install` _after_ `conda env create -f environment.yml`? Might work.)

I'm not able to test it properly unfortunately, but it does seems to install correctly for me when I do `conda env create -f environment.yml` (and running `poetry install` after that does not affect the install at all).

Anyways, since you are the one to be using this I will leave it to you to decide what makes more sense.

**Comment by engeir** [comment]
That's good! Could you see if this works now?

**Comment by sajidah-ahmed** [comment]
> That's good! Could you see if this works now?

It works! :) Thanks @engeir! I'll update the README.md with the new setup details.

File: pyproject.toml
Comment (gregordecristoforo) [inline pyproject.toml:19]: Please change version number to 0.1.2 so that we can upload the updates version to pip
Also update version number in __init__.py

**Review by gregordecristoforo — CHANGES_REQUESTED** [review]

**Comment by engeir** [comment]
Okay, then let's just stick with the working procedure. @sajidah-ahmed could you, since you are able to test it, fix the `README` with instructions and remove `environment.yml` and possibly other unneeded files?

**Review by gregordecristoforo — APPROVED** [review]

**Review by sajidah-ahmed — APPROVED** [review]

**Pull request merged**
